src/users/src/main/java/engineering/everest/lhotse/users/services/DefaultUsersService.java

```diff
@@ -1,45 +1,71 @@
 
-    public DefaultUsersService(HazelcastCommandGateway commandGateway,
-                               RandomFieldsGenerator randomFieldsGenerator,
-                               PasswordEncoder passwordEncoder) {
+    public DefaultUsersService(HazelcastCommandGateway commandGateway, KeycloakSynchronizationService keycloakSynchronizationService) {
         this.commandGateway = commandGateway;
-        this.randomFieldsGenerator = randomFieldsGenerator;
-        this.passwordEncoder = passwordEncoder;
+        this.keycloakSynchronizationService = keycloakSynchronizationService;
     }
 
     @Override
-    public void updateUser(UUID requestingUserId, UUID userId, String emailChange,
-                           String displayNameChange, String passwordChange) {
+    public void updateUser(UUID requestingUserId, UUID userId, String emailChange, String displayNameChange) {
+        commandGateway.sendAndWait(new UpdateUserDetailsCommand(userId, emailChange, displayNameChange, requestingUserId));
+    }
 
-        commandGateway.sendAndWait(new UpdateUserDetailsCommand(userId, emailChange,
-                displayNameChange, encodePasswordIfNotBlank(passwordChange), requestingUserId));
+    @Override
+    public void updateUserRoles(UUID requestingUserId, UUID userId, Set<Role> roles) {
+        commandGateway.sendAndWait(new UpdateUserRolesCommand(userId, roles, requestingUserId));
     }
 
     @Override
-    public UUID createUser(UUID requestingUserId, UUID organizationId, String username, String displayName, String rawPassword) {
-        return commandGateway.sendAndWait(new CreateUserCommand(randomFieldsGenerator.genRandomUUID(), organizationId, requestingUserId,
-                username, encodePasswordIfNotBlank(rawPassword), displayName));
+    public UUID createUser(UUID requestingUserId, UUID organizationId, String username, String displayName) {
+        try {
+            // Create user in the keycloak first so that we can get a newly created user id and map it in our app db.
+            keycloakSynchronizationService.createUser(
+                    Map.of("username", username,
+                            "email", username,
+                            "enabled", true,
+                            "attributes", new UserAttribute(organizationId, Set.of(Role.ORG_USER), "Guest"),
+                            "credentials", List.of(
+                                    Map.of("type", "password",
+                                            "value", "changeme",
+                                            "temporary", true))));
+
+            return commandGateway.sendAndWait(new CreateUserCommand(getUserId(username), organizationId,
+                    requestingUserId, username, displayName));
+        } catch (RemoteCommandExecutionException tex) {
+            LOGGER.info("Keycloak createUser error: " + tex);
+            return commandGateway.sendAndWait(new CreateUserCommand(getUserId(username), organizationId,
```

Comment:
We have introduced this to satisfy the `Domain Validation Error Messages Are Internationalized`. Otherwise it will cause one of the existing functional test cases to fail.

`createUser` method is expected to return an UUID. If an email already exists in the system, error message is expected. And the error message should be internationalized.

Currently we don't have the i18n support for keycloak related messages.

Reply:
This doesn't sound at all related to the command being dispatched as a way of handling the RemoteCommandExecutionException.

Let's talk about this today.

Reply:
Updated the exception handling.
Reverted back the command handling changes.
Keycloak exceptions are simply logged.